crates/flipperzero/src/furi/string.rs

```diff
@@ -0,0 +1,784 @@
+//! String primitives built around `FuriString`.
+
+use core::{
+    cmp::Ordering,
+    convert::Infallible,
+    ffi::{c_char, CStr},
+    fmt::{self, Write},
+    hash,
+    ops::{Add, AddAssign},
+    ptr,
+};
+
+#[cfg(feature = "alloc")]
+use alloc::{borrow::Cow, boxed::Box, ffi::CString};
+
+use flipperzero_sys as sys;
+
+mod pattern;
+use self::pattern::Pattern;
+
+/// Source: [UnicodeSet `[:White_Space=Yes:]`][src].
+///
+/// [src]: https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=[%3AWhite_Space%3DYes%3A]
+const WHITESPACE: &[char] = &[
+    ' ', '\x09', '\x0a', '\x0b', '\x0c', '\x0d', '\u{0085}', '\u{00A0}', '\u{1680}', '\u{2000}',
+    '\u{2001}', '\u{2002}', '\u{2003}', '\u{2004}', '\u{2005}', '\u{2006}', '\u{2007}', '\u{2008}',
+    '\u{2009}', '\u{200A}', '\u{2028}', '\u{2029}', '\u{202F}', '\u{205F}', '\u{3000}',
+];
+
+/// A Furi string.
+///
+/// This is similar to Rust's [`CString`] in that it represents an owned, C-compatible,
+/// nul-terminated string with no nul bytes in the middle. It also has additional methods
+/// to provide the flexibility of Rust's [`String`]. It is used in various APIs of the
+/// Flipper Zero SDK.
+///
+/// This type does not requre the `alloc` feature flag, because it does not use the Rust
+/// allocator. Very short strings (7 bytes or fewer) are stored directly inside the
+/// `FuriString` struct (which is stored on the heap), while longer strings are allocated
+/// on the heap by the Flipper Zero firmware.
+#[derive(Eq)]
+pub struct String(*mut sys::FuriString);
```

Comment:
I'm still wondering if we should explicitly call this `FuriString` or just `String`.

`String` is nice since, if we included it in the prelude, things almost look like regular Rust (`String::from("word")` does what you expect, common functions use `String`, etc.). However, I think this leads to more confusion in the long term (accidentally looking at the `std::string::String`  docs, wondering why a library API that takes a `String` doesn't work, not being clear whether an API takes a standard Rust `String` or a Furi `String`, etc.).

I'm leaning towards `FuriString` at the moment.

Reply:
Originally I was hoping for this to actually be a drop-in replacement for `String`. Now that it isn't the case, I agree with renaming it for clarity.